### ISSUE TITLE
Read gzipped SDDS files

### ISSUE THREAD

**Issue by chernals** [pylhc/sdds#57]
### Feature Description

Would be nice to be able to read directly gzipped files. A minor modification to the `read_sdds` function would do the trick, a bit of refactoring would allow to add it on top of a common interface for both functions.

### Possible Implementation

```python

def read_sdds_gz(file_path: Union[pathlib.Path, str], endianness: str = None) -> SddsFile:
    """
    Reads SDDS file from the specified ``file_path``.

    Args:
        file_path (Union[pathlib.Path, str]): `Path` object to the input SDDS file. Can be a
            `string`, in which case it will be cast to a `Path` object.
        endianness (str): Endianness of the file. Either 'big' or 'little'.
                          If not given, the endianness is either extracted from
                          the comments in the header of the file (if present)
                          or determined by the machine you are running on.
                          Binary files written by this package are all big-endian,
                          and contain a comment in the file.

    Returns:
        An `SddsFile` object containing the loaded data.
    """
    file_path = pathlib.Path(file_path)
    with gzip.open(file_path,'rb') as inbytes:
#     with file_path.open("rb") as inbytes:
        if endianness is None:
            endianness = _get_endianness(inbytes)
        version, definition_list, description, data = _read_header(inbytes)
        data_list = _read_data(data, definition_list, inbytes, endianness)

        return SddsFile(version, description, definition_list, data_list)
```

**Comment by chernals** [comment]
If you think that it makes sense I'm happy to submit a PR.

**Comment by JoschD** [comment]
Hey @chernals , thank you very much for your request. Should be easy enough to implement.

To the team:
I would suggest something more like passing an opener to the function.

```python
import os
import io
from contextlib import contextmanager, AbstractContextManager
from typing import Callable

OpenerType = Callable[[os.PathLike, str], AbstractContextManager[io.BufferedIOBase]]
   

def read_sdds(file_path: Union[pathlib.Path, str], endianness: str = None, opener: OpenerType = open) -> SddsFile:
    """
    Reads SDDS file from the specified ``file_path``.

    Args:
        file_path (Union[pathlib.Path, str]): `Path` object to the input SDDS file. Can be a
            `string`, in which case it will be cast to a `Path` object.
        endianness (str): Endianness of the file. Either 'big' or 'little'.
                          If not given, the endianness is either extracted from
                          the comments in the header of the file (if present)
                          or determined by the machine you are running on.
                          Binary files written by this package are all big-endian,
                          and contain a comment in the file.
        gz (bool): Flag to declare, that the file is in .gz format. It will be unpacked on-the-fly.

    Returns:
        An `SddsFile` object containing the loaded data.
    """
    file_path = pathlib.Path(file_path)   
    with opener(file_path, "rb") as inbytes:
        if endianness is None:
            endianness = _get_endianness(inbytes)
        version, definition_list, description, data = _read_header(inbytes)
        data_list = _read_data(data, definition_list, inbytes, endianness)

        return SddsFile(version, description, definition_list, data_list)
``` 

Opinions? @Mael-Le-Garrec @fsoubelet @fscarlier @awegsche

**Comment by chernals** [comment]
@JoschD Very nice, that's what I had in mind! Maybe you can also provide / expose a gzip opener (as you `binary_open`), for lazy users ;)

**Comment by fsoubelet** [comment]
I'm ok with this with a few points:
- I thing `typing.Callable` is deprecated now and we should use the equivalent from `collections`.
- This adds an abstraction layer for the user to manipulate so I would like to see quite some detail added in the documentation to guide them, with at least an example to showcase how to use another one.
- As @chernals said, it can be good to expose ourselves one or more alternative openers for the user to call without having to do tweaks themselves.

👍

**Comment by fsoubelet** [comment]
Closed with #58 

@chernals you can expect the new version to be available from PyPI in a few minutes.
